spec.md

```diff
@@ -1925,6 +1929,123 @@ To disallow infinitely recursive hardware, modules cannot contain instances of
 itself, either directly, or indirectly through instances of other modules it
 instantiates.
 
+## Optional Groups
+
+Optional groups are named collections of statements inside a module.  Optional
+groups contain optional functionality for a module that can be included or
+excluded without affecting the intended functionality of a circuit.  Optional
+groups are used to keep verification, debugging, or FIRRTL statements, not
+relevant to the operation of the circuit, in a separate area.  Optional groups
+have special emission rules defined in the FIRRTL ABI.
+
+The `defgroup`{.firrtl} statement declares an optional group with a specific
+name.  A `defgroup`{.firrtl} may declare an optional group for a
+`circuit`{.firrtl} or for an `extmodule`{.firrtl}.  The name of an optional
+group participate specially in the module namespace--no module may include a
+suffix of `_<group-name>` for any declared group name.
+
+The `group`{.firrtl} statement creates an optional group instance within a
+module.  A `group`{.firrtl} must refer to an optional group name declared by a
+`defgroup`{.firrtl} of the current `circuit`{.firrtl}.  A `group`{.firrtl}
+includes statements which refer to sources outside the group.
+
+The `circuit`{.firrtl} below contains one, `Bar`.  Module `Foo` contains a group
+instance that creates a node computed from a port defined outside its scope.
+
+``` firrtl
+circuit Foo:
+  defgroup Bar:          ; declare the group Bar
+
+  module Foo:
+    input a: UInt<1>
+
+    group bar of Bar:    ; create the group instance bar
+      node notA = not(a)
+
+  extmodule Baz:
+    input a: UInt<1>
+    output b: UInt<1>
+    defgroup Bar:        ; a group may be declared on an external module
+```
+
+Groups may be nested.  Nested groups are declared with the `defgroup`{.firrtl}
+statement indented under an existing `defgroup`{.firrtl}.  The circuit below
+contains four groups.  `Bar` is the top-level group.  `Baz` and `Qux` are nested
+under `Bar`.  `Quz` is nested under `Qux`.
+
+``` firrtl
+circuit Foo:
+  defgroup Bar:
```

Comment:
They are now opening a possibly empty scope. It's the same syntax as empty module (which is much rarer):

```
circuit Foo:
  module Bar:
  module Foo:
```

Maybe there's a better approach, e.g., `:` is optional if there is no scope or this could use 😨 `{}`.

I'd like to keep this consistent with how other scopes work (module, when, etc.) and then do a batch syntax change for all scopes.

Reply:
for some reason i thought you had to write `pass` for an empty module